### ISSUE TITLE
PSA on 1.9 Performance Regression

### ISSUE THREAD

**Issue by quinnj** [JuliaLang/julia#50762]
Was doing some performance investigation on 1.9, and noticed the following:

Setup:
```julia
using JSON3, BenchmarkTools
json = """
{
"store": {
    "book": [
    {
        "category": "reference",
        "author": "Nigel Rees",
        "title": "Sayings of the Century",
        "price": 8.95
    },
    {
        "category": "fiction",
        "author": "Herman Melville",
        "title": "Moby Dick",
        "isbn": "0-553-21311-3",
        "price": 8.99
    },
    {
        "category": "fiction",
        "author": "J.R.R. Tolkien",
        "title": "The Lord of the Rings",
        "isbn": "0-395-19395-8",
        "price": 22.99
    }
    ],
    "bicycle": {
    "color": "red",
    "price": 19.95
    }
},
"expensive": 10
}
"""
```

Then we get the following results:

1.8:
```julia
julia> @btime x[:store][:book]
  12.000 μs (14 allocations: 1.03 KiB)
```

vs. 1.9:
```julia
julia> @btime x[:store][:book]
  88.000 μs (55 allocations: 2.73 KiB)
```

then again on Julia master:
```julia
julia> @btime x[:store][:book]
  1.188 μs (14 allocations: 1.02 KiB)
```

Checking out a profile for 1.9, we see the following:
<img width="1342" alt="image" src="https://github.com/JuliaLang/julia/assets/2896623/0a5f1617-3599-426e-beec-2522fe7b412f">

Which seems to suggest this has something to do with sparams, which made me think back to https://github.com/JuliaLang/julia/issues/48229, but this seems slightly different, because things _do_ seem resolved on master. 

Anyway, I'm actually fine if we just close this, since things seem to be fixed on master, but I at least wanted to give a public report of my investigations in case it helps anyone else trying to track down 1.9 performance issues.

**Comment by KristofferC** [comment]
If someone is up for it it the fix could be bisected and maybe backported.

**Comment by gbaraldi** [comment]
Is it fixed at least on 1.10?

**Comment by quinnj** [comment]
Ah, let me check 1.10 specifically

**Comment by quinnj** [comment]
Welp, I get this from the `backports-release-1.10` branch, let me try `v1.10-beta1`

```julia
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.0-beta1.6 (2023-07-28)
 _/ |\__'_|_|_|\__'_|  |  backports-release-1.10/21fe72332b (fork: 51 commits, 30 days)
|__/                   |

julia> using JSON3, BenchmarkTools
ERROR: MethodError: no method matching precompile(::String; _from_loading::Bool)

Closest candidates are:
  precompile(::Union{AbstractString, Pkg.Types.PackageSpec}; kwargs...) (method too new to be called from this world context.)
   @ Pkg ~/juliax/usr/share/julia/stdlib/v1.10/Pkg/src/API.jl:146
  precompile(::Pkg.Types.Context, ::Vector{Pkg.Types.PackageSpec}; internal_call, strict, warn_loaded, already_instantiated, timing, _from_loading, kwargs...) (method too new to be called from this world context.)
   @ Pkg ~/juliax/usr/share/julia/stdlib/v1.10/Pkg/src/API.jl:1078
  precompile(; name, uuid, version, url, rev, path, mode, subdir, kwargs...) (method too new to be called from this world context.)
   @ Pkg ~/juliax/usr/share/julia/stdlib/v1.10/Pkg/src/API.jl:165
  ...

Stacktrace:
  [1] _require(pkg::Base.PkgId, env::String)
    @ Base ./loading.jl:1957
  [2] __require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base ./loading.jl:1806
...
```

**Comment by quinnj** [comment]
Hmmm, same error. @KristofferC, any idea what's going on there?

**Comment by quinnj** [comment]
Relatedly, found a way to avoid the problematic code path all together so things are even faster than before: https://github.com/quinnj/JSON3.jl/pull/267

**Comment by KristofferC** [comment]
I'll close this  for now since the regression has been worked around in the package for 1.9 and it seems fixed on 1.10.
